module/zfs/dmu_recv.c

```diff
@@ -1408,7 +1408,7 @@ do_corrective_recv(struct receive_writer_arg *rwa, struct drr_write *drrw,
 		void *buf = abd_to_buf(cabd);
 		uint64_t csize = zio_compress_data(BP_GET_COMPRESS(bp),
 		    abd, &buf, abd_get_size(abd),
-		    rwa->os->os_complevel);
+		    rwa->os->os_complevel, 0);
```

Comment:
won't this break corrective receive in the same way as e.g. triggering gang blocks when they didn't exist before does in the test suite, by emitting a compressed record where one might not have existed before?

Reply:
`BP_GET_COMPRESS(bp) != ZIO_COMPRESS_OFF` in condition checks that on-disk BP has compressed flag, so the block was successfully compressed before (even with previous thresholds). And we already don't look at `csize` here, so we'll already write it as is. Hope I didn't miss anything.

Reply:
@alek-p I'd be glad if you may have time to recheck me here as the author of corrective recv, thank you!

Reply:
I think I agree with you, I just am wary because I know there's already existing failure modes around corrective receive if the output is not JUUUST right, but reading more context, I don't see a way this could be wrong unless the compressor fails nondeterministically, which would break many things already.

Reply:
> unless the compressor fails nondeterministically, which would break many things already.

you reminded me that compressor code has fast fail back if it can't allocate memory, and this case may give us problems everywhere where we wait for idempotent compression results. But it's not a new problem, so just a thought for now.

Reply:
I _think_ there might be corner cases that could impact corrective recv functionality if we have an old system without this patch that is using a compressed/raw sendfile from a newer system with this patch for corrective recv.
I believe the workaround would be to generate a sendfile that is not compressed (not using "zfs send -c/-w")

Reply:
Older hosts I think would only break on -w, and you can't really do much about that. -c you could just not use -c or use zstream decompress as an input and make it redo it.